File: src/python/tests/event-test.py
import argparse
import glib
import appfw
import logging
import sys
logger = logging.getLogger("__event-test__")

#Create mainloop in default context
ml = glib.MainLoop(None)

def event_handler(event, data):
	logger.debug("Received an event with event = " + str(event) + \
	             ", data = " + str(data))
	if (event == 'system::terminate'):
		print "Received a SIGTERM, quitting mainloop..."
		ml.quit() #This is the proper way to quit
	elif (event == 'system::reload'):
		print "Received SIGHUP, doing nothing..."

def status_handler(seqno, status, msg, data, user_data):
	logger.debug("Received a status message with status = " + str(status) + \
	             ", msg = " + str(msg) + ", data = " + str(data) +
	             ", user_data " + str(user_data))
	if (status == 0):
		print "Succesfully set new event subscriptions."
		print "New subscriptions are: " + str(user_data.subscriptions)
	else:
		print "Event subscription failed. Is the appfw server running?"

def send_handler_burst(id, status, msg, callback_data):
	logger.debug("Sent a burst event with id = " + str(id) + \
	             ", status = " + str(status) + ", msg = " + str(msg) + \
	             ", callback_data = " + str(callback_data))

def send_handler_timed(id, status, msg, callback_data):
	logger.debug("Sent a timed event with id = " + str(id) + \
	             ", status = " + str(status) + ", msg = " + str(msg) + \
	             ", callback_data = " + str(callback_data))

def event_sender_factory(event, event_data, func, callback_data, **kwargs):
	def event_sender(app, count):
		if event_sender.counter >= count:
			return False
		else:
			app.send_event(event, event_data, func, callback_data, **kwargs)
			event_sender.counter += 1
			return True

	event_sender.counter=0
	return event_sender

def be_server(app, args):
	app.user_data = app #Allows listing of subscriptions from status handler
	#Set subscriptions, event callback and status callback
	app.subscriptions = args.events
	app.event_handler = event_handler
	app.status_handler = status_handler

	#Receive signals as events
	if (args.signals):
		app.enable_signals()

	ml.run()

def be_client(app, args):
	#Create timed event sender glib timeouts
	for event in args.events:
		sender_func = event_sender_factory(event, {"timed" : "event"}, send_handler_timed,
		                                   "timed event callback_data", \
										   label=args.label, appid=args.appid, \
										   binary=args.binary, user=args.user, \
										   process=args.process)
		glib.timeout_add(args.interval, sender_func, app, args.nevent)

	#Create burst sender glib timeouts
	for event in args.events:
		sender_func = event_sender_factory(event, ["this", "is", "an", "example", {"of" : "event data"}],
		                                   send_handler_burst, "burst event callback_data", \
		                                   label=args.label, appid=args.appid, \
										   binary=args.binary, user=args.user, \
										   process=args.process)
		glib.timeout_add(0, sender_func, app, args.burst)

	ml.run()

def main():
	#Parse arguments
	parser = argparse.ArgumentParser()
	parser.add_argument("-s", "--server", action="store_true", help="subscribe and wait for events")
	parser.add_argument("-l", "--label", type=str, help="target application label")
	parser.add_argument("-a", "--appid", type=str, help="target application id")
	parser.add_argument("-b", "--binary", type=str, help="target application binary path")
	parser.add_argument("-u", "--user", type=str, help="target application user")
	parser.add_argument("-p", "--process", type=int, help="target application process id")
	parser.add_argument("-e", "--events", type=str, help="events to send/subscribe for")
	parser.add_argument("-n", "--nevent", type=int, default=0, help="number of bundles to send")
	parser.add_argument("-S", "--signals", action="store_true", help="subscribe to system signals")
	parser.add_argument("-B", "--burst", type=int, default=0, help="number of events in initial burst")
	parser.add_argument("-I", "--interval", type=int, default=0, help=" between sending")
	parser.add_argument("-v", "--verbose", action="count", default=0, help="increase logging verbosity (ignored)")
	parser.add_argument("-d", "--debug", action="count", default=0, help="enable given debug configuration")
	args = parser.parse_args()
	args.events = args.events.split(',')

	#Create appfw context
	app = appfw.IotApp()

	if args.debug > 0:
		if args.debug >= 1:
			logging.basicConfig()
			logger.setLevel(logging.DEBUG)
		if args.debug >= 2:
			app._enable_debug([""])
		if args.debug >= 3:
			app._enable_debug(["@python-app-fw-wrapper.cpp"])
		if args.debug > 3:
			app._enable_debug(["*"])

	logger.debug(args)

	if args.server:
		be_server(app, args)
	else:
		be_client(app, args)

if __name__ == "__main__":
	main()

File: src/java/tests/EventTest.java
import ArgParser.*;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import com.intel.ostro.appfw.*;

public class EventTest {
    public static void main(String [] args) {
        ArgParser parser = new ArgParser(args);
        addCommands(parser);
        Map<String, String> arguments = parser.parse();
        if (arguments.isEmpty() || arguments.containsKey("help")) {
            parser.printHelp();
            return;
        }

        if (arguments.containsKey("debug")) {
            AppFw fw = AppFw.getInstance();
            switch (arguments.get("debug")) {
                case "all":
                    fw.enableDebug(Arrays.asList("*"));
                    break;
                case "limited":
                    fw.enableDebug(Arrays.asList("@AppFw.c"));
                    break;
                case "none":
                    break;
                default:
                    throw new IllegalArgumentException(
                        "Invalid argument for debug: " + arguments.get("debug"));
            }
        }


        commonInit(arguments);

        if (arguments.containsKey("server")) {
            beServer(arguments);
        } else {
            beClient(arguments);
        }
    }

    public static void addCommands(ArgParser parser) {
        parser.addCommand(new Command("help", "Prints help"));
        parser.addCommand(new Command("server", "Subscribe and wait for events"));
        parser.addCommand(new CommandWithArg("label", "Target application label"));
        parser.addCommand(new CommandWithArg("user", 
            "Target application user"));
        parser.addCommand(new CommandWithArg("appid", 
            "Target application appid"));
        parser.addCommand(new CommandWithArg("binary", 
            "Target application binary path"));
        parser.addCommand(new CommandWithArg("process", 
            "Target application process id"));
        parser.addCommand(new CommandWithArg("events", 
            "Comma separated list of events that will be subscribed or sent"));
        parser.addCommand(new CommandWithArg("nevent", "Number of bundles to send"));
        parser.addCommand(new CommandWithArg("debug", 
            "Turn debugging on (valid arguments are none, limited, all)"));
        parser.addCommand(new CommandWithArg("Interval", 
            "Interval between event bundles"));
        
    }

    public static void commonInit(Map<String, String> arguments) {
        if (!arguments.containsKey("events") 
            || arguments.get("events").split(",").length == 0) {
            throw new IllegalArgumentException("No event subscriptions were provided for the server");
        }

        AppFw fw = AppFw.getInstance();

        // clean up code
        Runtime.getRuntime().addShutdownHook(new Thread() {
        @Override
            public void run() {
                System.out.println("Releasing resources...");
                fw.stopMainLoop();
                fw.close();
                System.out.println("Done");
            }   
        });

        // callback for event notifications
        fw.setEventCallback((String event, String json, Object userData) -> {
            System.out.println("Event: " + event 
                + " json data: " + json
                + " user data: " + (String)userData);
        },
        "this is my optional user data");
        
        // callback for event subscription change notifications
        fw.setStatusCallback(
                (int id, 
                 int status, 
                 String msg, 
                 String json, 
                 Object userData) -> {
            System.out.println("Got status update with id " + id + " and json: " + json);
            if (status == 0) {
                System.out.println(
                    "Status OK, succesfully subscribed to events " 
                    + ((AppFw)userData).getEventSubscriptions());
            } else {
                System.out.println(
                    "An error occured while subscribing to events: " + msg);
            }
        }, fw.getInstance()); // pass as user data for demonstration purposes
    }

    public static void beServer(Map<String, String> arguments) {

        String [] events = arguments.get("events").split(",");

        AppFw fw = AppFw.getInstance();
        fw.setEventSubscriptions(new HashSet<String>(Arrays.asList(events)));        
        fw.startMainLoop();
    }


    public static void beClient(Map<String, String> arguments) {
        AppFw fw = AppFw.getInstance();

        int values = 0;

        TargetApplication targetApplication = new TargetApplication();

        if (arguments.containsKey("label")) {
            targetApplication.setLabel(arguments.get("label"));
            ++values;
        }

        if (arguments.containsKey("user")) {
            targetApplication.setUser(arguments.get("user"));
            ++values;
        }

        if (arguments.containsKey("appid")) {
            targetApplication.setAppID(arguments.get("appid"));
            ++values;
        }

        if (arguments.containsKey("binary")) {
            targetApplication.setBinary(arguments.get("binary"));
            ++values;
        }        

        if (arguments.containsKey("process")) {
            int processID = 0;
            try {
                processID = Integer.parseInt(arguments.get("process"));
            } catch (NumberFormatException ex) {
                throw new IllegalArgumentException(
                    "Process argument must be an integer");
            }
            targetApplication.setProcess(processID);
            ++values;
        }        

        if (values == 0) {
            throw new IllegalArgumentException("At least one of label, user, "
                + "appid, binary path or process id must be provided when "
                + "sending events");
        }

        sendIntervalEvents(arguments, targetApplication);
        

        fw.startMainLoop();
    }

    public static void sendIntervalEvents(Map<String, String> arguments, 
        final TargetApplication targetApplication) {
        int interval;
        int eventBundles = 0;
        try {
            interval = Integer.parseInt(arguments.get("Interval"));
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException("Interal argument must be an integer");
        }
        try {
            if (arguments.containsKey("nevent")) {
                eventBundles = Integer.parseInt(arguments.get("nevent"));
            } else {
                eventBundles = 1;
            }
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException("nevents argument must be an integer");
        }


        final int finalInterval = interval;
        final int finalEventBundles = eventBundles;
        final String [] events = arguments.get("events").split(",");
        // AppFw is not thread safe and native access must be synchronized
        new Thread(() -> {
            for (int i = finalEventBundles; i > 0; --i) {
                // at this point, fw has been created so this merely returns the reference
                AppFw fw = AppFw.getInstance();
                synchronized (fw) {
                    for (int j = 0; j < events.length; ++j) {
                        fw.sendEvent(
                            events[j], 
                            "{\"example_key\": \"Example value\"}",
                            targetApplication,
                            (int id, int status, String message, Object userData) -> {
                                System.out.println(
                                    "Event send callback called with status "
                                    + status + ", message " + message 
                                    + " and user data " + (String)userData);
                            },
                            "user_data_here"
                            );
                    }
                }

                try {
                    Thread.sleep(finalInterval);;
                } catch (InterruptedException ex) {
                    System.out.println("Interval sleep was interrupted...");
                }
            }
            System.out.println("Interval sending finished");
        }).start();
    }
}

